lib/Pipes/TaggedFunctionKind.cpp

```diff
@@ -49,18 +55,223 @@ TaggedFunctionKind::symbolToTarget(const llvm::Function &Symbol) const {
+                          .Related;
+  for (const yield::RelationTarget &RT : Related) {
+    if (RT.Kind == yield::RelationType::IsDynamicallyCalledFrom) {
+      std::optional<MetaAddress>
+        Caller = genericLocationFromString<0>(RT.Location,
+                                              ranks::Function,
+                                              ranks::BasicBlock,
+                                              ranks::Instruction);
+      revng_assert(Caller.has_value());
+      Result.insert(Caller->toString());
+    }
+  }
+}
+
+using IsolatedFK = IsolatedFunctionKind;
+void IsolatedFK::getInvalidations(const Context &Ctx,
+                                  TargetsList &ToRemove,
+                                  const GlobalTupleTreeDiff &Diff,
+                                  const pipeline::Global &Before,
+                                  const pipeline::Global &After) const {
+  using PM = PathMatcher;
+  using Binary = model::Binary;
+  if (not llvm::isa<TupleTreeGlobal<Binary>>(Before))
+    return;
+
+  const auto &OldModel = cast<TupleTreeGlobal<Binary>>(Before).get();
+  const auto &Model = cast<TupleTreeGlobal<Binary>>(After).get();
+
+  // If CrossRelations is not found, the most conservative approach is to
+  // invalidate all targets.
+  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole)) {
+    ToRemove.emplace_back(Target(*this));
+    return;
+  }
+
+  static constexpr auto &BCRR = BinaryCrossRelationsRole;
+  using CRContainer = TupleTreeContainer<CrossRelations>;
+  const auto &Container = Ctx.getReadOnlyContainer<CRContainer>(BCRR);
+  const TupleTree<CrossRelations> &CrossRelations = Container.get();
+
+  const TupleTreeDiff<Binary> *ModelDiff = Diff.getAs<Binary>();
+  if (not ModelDiff)
+    return;
+
+  auto AttrMatcher = PM::create<Binary>("/Functions/*/Attributes").value();
+
+  std::set<std::string> Targets;
+  const auto &IDF = Model->ImportedDynamicFunctions;
 
-  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole))
+  for (const Change<Binary> &Change : ModelDiff->Changes) {
+    revng_assert(not Change.Path.empty());
+
+    // A new function has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Functions") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
+      } else if (Change.Old) {
+        // Invalidate the function and its callers
+        auto Function = std::get<model::Function>(*Change.Old);
+        Targets.insert(Function.Entry.toString());
+        insertCallers(CrossRelations, Function.Entry, Targets);
+      }
+    }
+
+    // A new prototype has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Types") {
+      if (Change.New) {
+        // Invalidate all targets
+        for (const model::Function &Function : Model->Functions)
+          Targets.insert(Function.Entry.toString());
```

Comment:
Why should we invalidate all the functions when a new type is added?

Reply:
Indeed, we do not have too. Removed that part and kept it for `ABIEnforced` Kind, if anything.

Reply:
Uhm, no, I don't see why adding a new type should trigger an invalidation in general.